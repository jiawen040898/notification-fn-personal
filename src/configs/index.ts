export * from './aws.config';
export * from './pusher.config';
