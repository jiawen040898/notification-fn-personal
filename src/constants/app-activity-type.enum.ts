export enum AppActivityType {
    EMPLOYEE_TASK_ASSIGNED = 'employee_task_assigned',
    EMPLOYEE_GOAL_ASSIGNED = 'employee_goal_assigned',
    EMPLOYEE_COURSE_ASSIGNED = 'employee_course_assigned',
    EMPLOYEE_ASSIGNMENT_DUE = 'employee_assignment_due',
    EMPLOYEE_ASSIGNMENT_OVERDUE = 'employee_assignment_overdue',
    EMPLOYEE_GOAL_MARK_COMPLETED = 'employee_goal_mark_completed',
    EMPLOYEE_COURSE_MARK_COMPLETED = 'employee_course_mark_completed',
    EMPLOYEE_TASK_MARK_COMPLETED = 'employee_task_mark_completed',
    EMPLOYEE_GOAL_UPDATED = 'employee_goal_updated',
    EMPLOYEE_COURSE_UPDATED = 'employee_course_updated',
    EMPLOYEE_TASK_UPDATED = 'employee_task_updated',
    SUBORDINATE_GOAL_UPDATED = 'subordinate_goal_updated',
    SUBORDINATE_COURSE_UPDATED = 'subordinate_course_updated',
    SUBORDINATE_TASK_UPDATED = 'subordinate_task_updated',
    EMPLOYEE_GOAL_DELETED = 'employee_goal_deleted',
    EMPLOYEE_COURSE_DELETED = 'employee_course_deleted',
    EMPLOYEE_TASK_DELETED = 'employee_task_deleted',
    EMPLOYEE_FEEDBACK_NOMINATION_REQUESTED = 'employee_feedback_nomination_requested',
    EMPLOYEE_FEEDBACK_RECOMMENDATION_READY = 'employee_feedback_recommendation_ready',
    EMPLOYEE_FEEDBACK_NOMINATION_SUBMISSION_DUE = 'employee_feedback_nomination_due',
    MANAGER_FEEDBACK_NOMINATION_APPROVAL_REQUESTED = 'manager_feedback_nomination_approval_requested',
    MANAGER_FEEDBACK_NOMINATION_APPROVAL_DUE = 'manager_feedback_nomination_approval_due',
    EMPLOYEE_FEEDBACK_NOMINATION_APPROVED = 'employee_feedback_nomination_approved',
    EMPLOYEE_FEEDBACK_NOMINATION_AMENDMENT_APPROVED = 'employee_feedback_nomination_amendment_approved',
    EMPLOYEE_FEEDBACK_NOMINATION_ADMIN_AMENDMENT = 'employee_feedback_nomination_admin_amendment',
    MANAGER_FEEDBACK_NOMINATION_AUTO_APPROVED = 'manager_feedback_nomination_auto_approved',
    MANAGER_FEEDBACK_NOMINATION_ADMIN_AMENDMENT = 'manager_feedback_nomination_admin_amendment',
    EMPLOYEE_FEEDBACK_REVIEWEE_RESPONSE_REQUESTED = 'employee_feedback_reviewee_response_requested',
    EMPLOYEE_FEEDBACK_REVIEWER_RESPONSE_REQUESTED = 'employee_feedback_reviewer_response_requested',
    EMPLOYEE_FEEDBACK_REVIEWEE_SUBMISSION_DUE = 'employee_feedback_reviewee_submission_due',
    EMPLOYEE_FEEDBACK_REVIEWER_SUBMISSION_DUE = 'employee_feedback_reviewer_submission_due',
    EMPLOYEE_FEEDBACK_RESULT_READY = 'employee_feedback_result_ready',
    MANAGER_FEEDBACK_RESULT_DIRECT_REPORT_READY = 'manager_feedback_result_direct_report_ready',
    EMPLOYEE_PROGRAM_INVITATION = 'employee_program_invitation',
    EMPLOYEE_PROGRAM_RESULT_READY = 'employee_program_result_ready',
}
