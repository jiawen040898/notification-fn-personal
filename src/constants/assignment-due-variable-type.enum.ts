export enum AssignmentDueVariableType {
    TODAY = 'today',
    IN_3_DAYS = 'in 3 days',
}
