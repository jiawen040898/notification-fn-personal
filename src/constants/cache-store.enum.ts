export enum CacheObject {
    APP_ACTIVITY_MESSAGE = 'NOTIFICATION:APP_ACTIVITY_MESSAGE',
    USER_ACCOUNT_ID = 'USER_ACCOUNT_ID',
    COMPANY_ID = 'COMPANY:COMPANY_ID',
}
