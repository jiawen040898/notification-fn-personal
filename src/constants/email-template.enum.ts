export enum EmailTemplateCommunicationType {
    CONSOLE_ACCOUNT_INVITATION = 'console_account_invitation', //system
    ACCOUNT_INVITATION = 'account_invitation', //system ,new user
    WELCOME_ACCOUNT = 'welcome_account', //system, existing user
    EMPLOYEE_ACCOUNT_INVITATION = 'employee_account_invitation', //system, new user
    EMPLOYEE_WELCOME_ACCOUNT = 'employee_welcome_account', //system, existing user
    DATA_DOWNLOAD_CONFIRMATION = 'data_download_confirmation', //system, for candidate data export
    APPLICATION_DOCUMENT_DOWNLOAD_CONFIRMATION = 'application_document_download_confirmation', //system, for export candidate assessment report, resume/attachment
    PROGRAM_REMINDER = 'program_reminder', //employee
    PROGRAM_INVITATION = 'program_invitation', //employee
    PROGRAM_RESULT_READY = 'program_result_ready', //employee
    APPLICATION_INVITATION = 'application_invitation', //candidate, replace job_apply_invitation
    APPLICATION_REMINDER = 'application_reminder', //candidate
    APPLICATION_SHORTLIST = 'application_shortlist', //candidate
    APPLICATION_REJECT = 'application_reject', //candidate
    ASSESSMENT_INVITATION = 'assessment_invitation', //candidate
    VIDEO_INTERVIEW_INVITATION = 'video_interview_invitation', //candidate replace video_invitation
    VIDEO_INTERVIEW_REMINDER = 'video_interview_reminder', //candidate
    EMPLOYEE_ASSIGNMENT_NOTIFICATION = 'employee_assignment_notification', // employee
    EMPLOYEE_ASSIGNMENT_REMINDER_NOTIFICATION = 'employee_assignment_reminder_notification', // employee
    EMPLOYEE_ASSIGNMENT_COMPLETION_NOTIFICATION = 'employee_assignment_completion_notification', // employee
    EMPLOYEE_ASSIGNMENT_UPDATED_NOTIFICATION = 'employee_assignment_updated_notification', // employee
    EMPLOYEE_ASSIGNMENT_DELETED_NOTIFICATION = 'employee_assignment_deleted_notification', // employee
    MANAGER_SUBORDINATE_PROGRESS_REPORT = 'manager_subordinate_progress_report', // employee

    EMPLOYEE_FEEDBACK_NOMINATION_REQUESTED = 'employee_feedback_nomination_requested',
    MANAGER_FEEDBACK_NOMINATION_APPROVAL_REQUESTED = 'manager_feedback_nomination_approval_requested',
    EMPLOYEE_FEEDBACK_NOMINATION_SUBMISSION_REMINDER = 'employee_feedback_nomination_submission_reminder',
    MANAGER_FEEDBACK_NOMINATION_APPROVAL_REMINDER = 'manager_feedback_nomination_approval_reminder',
    EMPLOYEE_FEEDBACK_REVIEWEE_RESPONSE_SUBMISSION_REMINDER = 'employee_feedback_reviewee_response_submission_reminder',
    EMPLOYEE_FEEDBACK_REVIEWER_RESPONSE_SUBMISSION_REMINDER = 'employee_feedback_reviewer_response_submission_reminder',
    EMPLOYEE_FEEDBACK_NOMINATION_APPROVED = 'employee_feedback_nomination_approved',
    EMPLOYEE_FEEDBACK_NOMINATION_AMENDMENT_APPROVED = 'employee_feedback_nomination_amendment_approved',
    MANAGER_FEEDBACK_NOMINATION_AUTO_APPROVED = 'manager_feedback_nomination_auto_approved',
    EMPLOYEE_FEEDBACK_REVIEWEE_RESPONSE_REQUESTED = 'employee_feedback_reviewee_response_requested',
    EMPLOYEE_FEEDBACK_REVIEWER_RESPONSE_REQUESTED = 'employee_feedback_reviewer_response_requested',
    EMPLOYEE_FEEDBACK_RESULT_READY = 'employee_feedback_result_ready',
    MANAGER_FEEDBACK_RESULT_DIRECT_REPORT_READY = 'manager_feedback_result_direct_report_ready',
    EMPLOYEE_FEEDBACK_NOMINATION_ADMIN_AMENDMENT = 'employee_feedback_nomination_admin_amendment',
    MANAGER_FEEDBACK_NOMINATION_ADMIN_AMENDMENT = 'manager_feedback_nomination_admin_amendment',
    EMPLOYEE_FEEDBACK_RECOMMENDATION_READY = 'employee_feedback_recommendation_ready',
}

export enum EmailTemplateUsage {
    SYSTEM = 'system',
    DEFAULT = 'default',
    CLIENT = 'client',
}

export const eventToEmailTemplate = {
    talent_management_task_assigned: [
        EmailTemplateCommunicationType.EMPLOYEE_ASSIGNMENT_NOTIFICATION,
    ],
    talent_management_goal_assigned: [
        EmailTemplateCommunicationType.EMPLOYEE_ASSIGNMENT_NOTIFICATION,
    ],
};
