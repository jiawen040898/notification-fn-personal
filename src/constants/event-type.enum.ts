export enum EventType {
    COMPANY_CREATED = 'company_created',
    APPLICATION_WITHDRAW = 'application_withdraw',
    TALENT_MANAGEMENT_TASK_ASSIGNED = 'talent_management_task_assigned',
    TALENT_MANAGEMENT_GOAL_ASSIGNED = 'talent_management_goal_assigned',
    TALENT_MANAGEMENT_GOAL_COMPLETED = 'talent_management_goal_completed',
    TALENT_MANAGEMENT_TASK_COMPLETED = 'talent_management_task_completed',
    TALENT_MANAGEMENT_GOAL_UPDATED = 'talent_management_goal_updated',
    TALENT_MANAGEMENT_TASK_UPDATED = 'talent_management_task_updated',
    TALENT_MANAGEMENT_GOAL_DELETED = 'talent_management_goal_deleted',
    TALENT_MANAGEMENT_TASK_DELETED = 'talent_management_task_deleted',
    TALENT_MANAGEMENT_EMPLOYEE_REMINDER = 'talent_management_employee_reminder',
    MANAGER_WEEKLY_SUBORDINATE_PROGRESS_REPORT = 'talent_management_manager_weekly_subordinate_progress_report',
    EMPLOYEE_FEEDBACK_NOMINATION_REQUESTED = 'talent_management_feedback_nomination_requested',
    MANAGER_FEEDBACK_NOMINATION_APPROVAL_REQUESTED = 'talent_management_feedback_manager_nomination_approval_requested',
    FEEDBACK_NOMINATION_SUBMISSION_DUE = 'talent_management_feedback_nomination_submission_due',
    FEEDBACK_MANAGER_NOMINATION_APPROVAL_DUE = 'talent_management_feedback_manager_nomination_approval_due',
    FEEDBACK_REVIEWEE_RESPONSE_SUBMISSION_DUE = 'talent_management_feedback_reviewee_response_submission_due',
    FEEDBACK_REVIEWER_RESPONSE_SUBMISSION_DUE = 'talent_management_feedback_reviewer_response_submission_due',
    FEEDBACK_NOMINATION_AMENDMENT_APPROVED = 'talent_management_feedback_nomination_amendment_approved',
    FEEDBACK_NOMINATION_ADMIN_AMENDMENT = 'talent_management_feedback_nomination_admin_amendment',
    FEEDBACK_NOMINATION_APPROVED = 'talent_management_feedback_nomination_approved',
    FEEDBACK_MANAGER_NOMINATION_AUTO_APPROVED = 'talent_management_feedback_manager_nomination_auto_approved',
    FEEDBACK_MANAGER_NOMINATION_ADMIN_AMENDMENT = 'talent_management_feedback_manager_nomination_admin_amendment',
    FEEDBACK_REVIEWEE_RESPONSE_REQUESTED = 'talent_management_feedback_reviewee_response_requested',
    FEEDBACK_REVIEWER_RESPONSE_REQUESTED = 'talent_management_feedback_reviewer_response_requested',
    FEEDBACK_REVIEWEE_RESULT_READY = 'talent_management_feedback_reviewee_result_ready',
    FEEDBACK_MANAGER_DIRECT_REPORT_RESULT_READY = 'talent_management_feedback_manager_direct_report_result_ready',
    FEEDBACK_REVIEWEE_FEEDBACK_RECOMMENDATION_COMPLETED = 'talent_management_feedback_reviewee_recommendation_completed',
    PARTICIPANT_PROGRAM_INVITATION = 'talent_management_employee_program_invitation',
    PARTICIPANT_PROGRAM_RESULT_READY = 'talent_management_employee_program_result_ready',
}
