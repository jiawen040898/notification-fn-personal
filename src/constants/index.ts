export * from './app-activity-type.enum';
export * from './assignment-due-variable-type.enum';
export * from './cache-store.enum';
export * from './email-template.enum';
export * from './event-type.enum';
export * from './icon-type.enum';
export * from './locale.enum';
