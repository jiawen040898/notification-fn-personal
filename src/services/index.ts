export * from './app-notification.service';
export * from './company.service';
export * from './company-email-template.service';
export * from './email-notification.service';
export * from './email-tracking.service';
export * from './identity.service';
export * from './push-notification.service';
export * from './send-email.service';
export * from './talent-management-notification.service';
