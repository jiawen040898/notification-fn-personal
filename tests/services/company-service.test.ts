import { cacheService, logger } from '@pulsifi/fn';
import axios from 'axios';

import { DefaultLocale, Locale } from '../../src/constants';
import { CompanyService } from '../../src/services';
import { testData } from './fixtures/company-service.test-data';

jest.mock('axios');
jest.mock('@pulsifi/fn/utils/logger.util');
jest.mock('@pulsifi/fn/services/cache.service');

describe('CompanyService', () => {
    let sut: CompanyService;

    beforeAll(async () => {
        sut = new CompanyService();
    });

    describe('getCompanyData', () => {
        beforeEach(() => {
            jest.clearAllMocks();
        });

        it('should return null if invalid company id', async () => {
            const actual = await sut.getCompanyData(testData.invalidCompanyId);

            expect(actual).toBeNull();
        });

        it('should return cached data if available', async () => {
            // Arrange
            jest.mocked(cacheService.get).mockResolvedValueOnce(
                testData.companyData,
            );

            // Act
            const actual = await sut.getCompanyData(testData.validCompanyId);

            // Assert
            expect(actual).toMatchSnapshot();
            expect(axios.get).not.toHaveBeenCalled();
        });

        it('should return expected data from api', async () => {
            // Arrange
            const axiosGetMock = jest.mocked(axios.get);
            axiosGetMock.mockResolvedValueOnce(testData.companyAxiosResponse);

            // Act
            const actual = await sut.getCompanyData(testData.validCompanyId);

            // Assert
            expect(actual).toMatchSnapshot();
            expect(axiosGetMock.mock.calls).toMatchSnapshot('axios.get');
        });

        it('should log error if error occurs', async () => {
            // Arrange
            const axiosGetMock = jest.mocked(axios.get);
            axiosGetMock.mockRejectedValueOnce(new Error('AxiosError'));

            // Act
            const action = () => sut.getCompanyData(testData.validCompanyId);

            // Assert
            await expect(action()).rejects.toThrowError();
            expect(jest.mocked(logger.error).mock.calls).toMatchSnapshot(
                'logger.error',
            );
        });
    });

    describe('getCompanyLocaleByCompanyData', () => {
        beforeEach(() => {
            jest.clearAllMocks();
        });

        it('should return default locale when company data is null', async () => {
            // Arrange
            jest.mocked(cacheService.get).mockResolvedValueOnce(null);

            // Act
            const actual = await sut.getCompanyLocaleByCompanyId(
                testData.validCompanyId,
            );

            // Assert
            expect(actual).toEqual(DefaultLocale);
        });

        it('should return default locale when company data has no default locale', async () => {
            // Arrange
            jest.mocked(cacheService.get).mockResolvedValueOnce(
                testData.companyDataThatHasNoDefaultLocale,
            );

            // Act
            const actual = await sut.getCompanyLocaleByCompanyId(
                testData.validCompanyId,
            );

            // Assert
            expect(actual).toEqual(DefaultLocale);
        });

        it('should return company locale when company data has a default locale', async () => {
            // Arrange
            jest.mocked(cacheService.get).mockResolvedValueOnce(
                testData.companyData,
            );

            // Act
            const actual = await sut.getCompanyLocaleByCompanyId(
                testData.validCompanyId,
            );

            // Assert
            expect(actual).toEqual(Locale.PT_BR);
        });
    });
});
