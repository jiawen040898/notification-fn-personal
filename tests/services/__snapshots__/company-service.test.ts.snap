// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanyService getCompanyData should log error if error occurs: logger.error 1`] = `
[
  [
    "Fail to getCompanyData",
    {
      "err": [Error: AxiosError],
      "url": "https://alb.test.pulsifi.me/identity/v1.0/public/companies/123",
    },
  ],
]
`;

exports[`CompanyService getCompanyData should return cached data if available 1`] = `
{
  "id": 123,
  "locales": [
    {
      "is_default": false,
      "locale": "en-US",
    },
    {
      "is_default": true,
      "locale": "pt-BR",
    },
  ],
  "name": "test company",
  "slug": "test-company",
  "timezone": "Asia/Kuala_Lumpur",
}
`;

exports[`CompanyService getCompanyData should return expected data from api 1`] = `
{
  "id": 123,
  "locales": [
    {
      "is_default": false,
      "locale": "en-US",
    },
    {
      "is_default": true,
      "locale": "pt-BR",
    },
  ],
  "name": "test company",
  "slug": "test-company",
  "timezone": "Asia/Kuala_Lumpur",
}
`;

exports[`CompanyService getCompanyData should return expected data from api: axios.get 1`] = `
[
  [
    "https://alb.test.pulsifi.me/identity/v1.0/public/companies/123",
  ],
]
`;
